caas/kubernetes/provider/constraints.go

```diff
@@ -23,5 +27,80 @@ var unsupportedConstraints = []string{
 func (k *kubernetesClient) ConstraintsValidator(ctx context.ProviderCallContext) (constraints.Validator, error) {
 	validator := constraints.NewValidator()
 	validator.RegisterUnsupported(unsupportedConstraints)
-	return validator, nil
+	return &constraintsValidator{Validator: validator}, nil
+}
+
+type constraintsValidator struct {
+	constraints.Validator
+}
+
+func validateTopologyKeyTagFunc(key string, topologySpreadKey string, tagList []string) error {
+	var topologySpreadValue, otherValue *string
+	// Search for the topologySpreadValue for topologyKeyTag
+	for tag := range tagList {
+		if strings.HasPrefix(tagList[tag], topologySpreadKey) {
+			trimmedValue := strings.TrimPrefix(tagList[tag], topologySpreadKey+"=")
+			topologySpreadValue = &trimmedValue
+			break
+		}
+	}
+	if topologySpreadValue == nil {
+		// Nothing to do, continue to the next check
+		return nil
+	}
+
+	for tag := range tagList {
+		if strings.HasPrefix(tagList[tag], key) {
+			trimmedValue := strings.TrimPrefix(tagList[tag], key+"=")
+			otherValue = &trimmedValue
+			break
+		}
+	}
+
+	if otherValue != nil && *topologySpreadValue == *otherValue {
+		return errors.New("podAffinity/antiPodAffinity's topology-key and topologySpread's cannot have the same value")
+	}
+	return nil
+}
+
+// Validate returns an error if the given constraints are not valid, and also
+// any unsupported attributes.
+func (v *constraintsValidator) Validate(cons constraints.Value) ([]string, error) {
+	if cons.Tags == nil {
+		return nil, nil
+	}
+	positives, negatives := parseDelimitedValues(*cons.Tags)
+	elements := append(positives, negatives...)
+
+	topologySpreadKey := caasApplication.TopologySpreadPrefix + caasApplication.TopologyKeyTag
+	podKey := caasApplication.PodPrefix + caasApplication.TopologyKeyTag
+	if err := validateTopologyKeyTagFunc(podKey, topologySpreadKey, elements); err != nil {
+		return nil, err
+	}
+
+	antiPodKey := caasApplication.AntiPodPrefix + caasApplication.TopologyKeyTag
+	if err := validateTopologyKeyTagFunc(antiPodKey, topologySpreadKey, elements); err != nil {
+		return nil, err
+	}
+
+	return v.Validator.Validate(cons)
```

Comment:
Do this check first and then the affinity checks. Avoids the awkward bug noted above.